Provide Bearer Token resolving strategy

vpavic opened the pull request:
This is a resubmit of spring-projects/spring-security#5122 which resolves spring-projects/spring-security#5121.

jzheaux commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/resource/web/DefaultBearerTokenResolver.java:
```diff
@@ -0,0 +1,93 @@
+ */
+public final class DefaultBearerTokenResolver implements BearerTokenResolver {
+
+	private static final Pattern authorizationPattern = Pattern.compile("^Bearer (?<token>[^\\s]+)*$");
```
Maybe this should check for base64 strings instead of simply checking for no whitespace, i.e. prefer white-list to black-list regexes.

vpavic replied:
Agreed. I planned to improve regex before submitting but forgot to do so.

jzheaux commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/resource/web/DefaultBearerTokenResolver.java:
```diff
@@ -0,0 +1,93 @@
+	}
+
+	private boolean isParameterTokenSupportedForRequest(HttpServletRequest request) {
+		return ((this.useFormEncodedBodyParameter && "POST".equals(request.getMethod()))
```
In the spec, it states that clients must not use this method unless

> The HTTP request entity-header includes the "Content-Type" header field set to "application/x-www-form-urlencoded"

One could argue that this is implied by it being a POST and having been parsed successfully by the servlet API. That said, it seems like a simple defensive measure to take here.

vpavic replied:
> this is implied by it being a POST and having been parsed successfully by the servlet API

That's exactly the reason why I omitted it. Checking the content type here is needless.

jzheaux replied:
Sounds great.

jzheaux commented on oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/resource/web/DefaultBearerTokenResolverTests.java:
```diff
@@ -0,0 +1,142 @@
+	}
+
+	@Test
+	public void resolveWhenFormParameterIsPresentAndSupportedThenTokenIsNotResolved() {
```
Should this be `ThenTokenIsResolved`? Same question for two tests down.

vpavic replied:
Its possible some test have typos in names due to c&p. Will review and amend.

jgrandja commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/resource/web/DefaultBearerTokenResolver.java:
```diff
@@ -0,0 +1,93 @@
+ * @since 5.1
+ * @see <a href="https://tools.ietf.org/html/rfc6750#section-2" target="_blank">RFC 6750 Section 2: Authenticated Requests</a>
+ */
+public final class DefaultBearerTokenResolver implements BearerTokenResolver {
```
I would prefer that we have a separate implementation for each resolving strategy. For example, the `DefaultBearerTokenResolver` becomes `AuthorizationHeaderBearerTokenResolver` and this is the only implementation we provide for now.

Later on we can add `RequestBodyBearerTokenResolver` and `QueryParameterBearerTokenResolver` implementation if need be. NOTE: Both of these are `optional` as per spec and I doubt we'll provide a `QueryParameterBearerTokenResolver` implementation but the user will have the ability to provide their own and configure it.

vpavic replied:
I don't believe that would be a good idea since we would be providing `BearerTokenResolver` implementations that violate the spec. The [RFC 6749 Section 2.1](https://tools.ietf.org/html/rfc6750#section-2.1) clearly states that Authorization header method:

> Clients SHOULD make authenticated requests with a bearer token using
   the "Authorization" request header field with the "Bearer" HTTP
   authorization scheme.  Resource servers MUST support this method.

Other methods can be supported on top, not instead of Authorization header method so the proposed implementation of `DefaultBearerTokenResolver` closely aligns with the spec.

jgrandja replied:
I think we are talking the same thing. To re-clarify, we provide one implementation of `BearerTokenResolver` only, this being the Authorization Request Header as it's required per spec. However, instead of naming it `DefaultBearerTokenResolver`, we rename it to be more specific to the strategy `AuthorizationHeaderBearerTokenResolver`.

vpavic replied:
I'm not following. In the previous comment, you mentioned adding `RequestBodyBearerTokenResolver` and `QueryParameterBearerTokenResolver` later on - those two clearly don't support the Authorization header method which is required by the spec.

I'd again like to stress that other token resolution methods are be supported on top of required Authorization header method, rather than replacing it as that would render resource server not compliant with the spec.

I also think we really should offer optional support of other methods as the effort is minimal, the old `spring-security-oauth2` also has that, and if you're building a resource server that's consumed by a variety of clients (i.e. not all are under your control) it's likely there will be older clients which use other methods.

jzheaux replied:
@jgrandja, if we rename it to `AuthorizationHeaderBearerTokenResolver`, are you also saying to remove the support already in the class for the other two strategies? The name seems a bit odd if we are leaving that functionality in.

And then, are you also suggesting we not provide support (initially) for the other two strategies?

> Other methods can be supported on top, not instead of Authorization header method

@vpavic, if I'm not mistaken, @jgrandja  is not suggesting that `RequestBodyBearerTokenResolver` would only resolve via the request body, but that it would first delegate to `AuthoriationHeaderBearerTokenResolver`, still then complying with the spec. I believe he is simply stating that the configuration you are going for can be achieved via a specialized class instead of via a flag.

For me, there is value in both of these approaches, the `Default` approach concisely addresses the spec all at once. The `AuthorizationHeader` approach is impossible to accidentally misconfigure--i.e. to get `RequestBody` support, a dev would need to change out the implementation, not just get a configuration flag incorrect.

vpavic replied:
> is not suggesting that `RequestBodyBearerTokenResolver` would only resolve via the request body, but that it would first delegate to `AuthoriationHeaderBearerTokenResolver`

It's not clear to me how would you make things work with this approach if you want to enable both  optional methods (form post body param & query param)? Also keep in mind that the spec states that _Clients MUST NOT use more than one method to transmit the token in each request_ and that `invalid_request` should be returned if they do. I don't see an easy way how to meet this requirement with different `BearerTokenResolver` without introducing significant amount of complexity.

> The `AuthorizationHeader` approach is impossible to accidentally misconfigure--i.e. to get `RequestBody` support, a dev would need to change out the implementation, not just get a configuration flag incorrect.

If currently proposed `DefaultBearerTokenResolver` is registered by default, to enable optionally supported methods one would need explicitly configure and register their own `DefaultBearerTokenResolver` instance - IMO that's pretty robust as well and shouldn't be easy to misconfigure by accident.

jzheaux replied:
> If currently proposed `DefaultBearerTokenResolver` is registered by default, to enable optionally supported methods one would need explicitly configure and register their own `DefaultBearerTokenResolver` instance.

I'll defer to @jgrandja on this point: Is default config sufficient coverage against accidental misconfiguration?

jgrandja replied:
@jzheaux  
> if we rename it to AuthorizationHeaderBearerTokenResolver, are you also saying to remove the support already in the class for the other two strategies? The name seems a bit odd if we are leaving that functionality in.

Correct, we remove the support for the other two strategies for `AuthorizationHeaderBearerTokenResolver`.

> And then, are you also suggesting we not provide support (initially) for the other two strategies?

I feel that we should focus our initial effort on `AuthorizationHeaderBearerTokenResolver` (the required one) and log a separate ticket for Request Body strategy and deal with that later. The Uri query parameter strategy has plenty of security deficiencies that I don't feel we should provide an implementation for it.

@vpavic 

> It's not clear to me how would you make things work with this approach if you want to enable both optional methods (form post body param & query param)...

I obviously did not do a good job communicating my points across so I figure code snippets will be the most clear. This is what I'm thinking...


```
public interface BearerTokenResolver {
}

// Required implementation
public class AuthorizationHeaderBearerTokenResolver implements BearerTokenResolver {
}

// Optional implementation
public class RequestBodyBearerTokenResolver implements BearerTokenResolver {
}

// Optional implementation NOT RECOMMENED
public class QueryParameterBearerTokenResolver implements BearerTokenResolver {
}

public class DelegatingBearerTokenResolver implements BearerTokenResolver {
	// Required as per spec
	private BearerTokenResolver defaultBearerTokenResolver = new AuthorizationHeaderBearerTokenResolver();

	// Optional as per spec
	private Set<BearerTokenResolver> bearerTokenResolvers;

	@Override
	public String resolve(HttpServletRequest request) {
		// Delegate to bearerTokenResolvers (if available)

		// If bearer token IS NOT resolved from (optional) bearerTokenResolvers than delegate to defaultBearerTokenResolver

		// If bearer token IS resolved from (optional) bearerTokenResolvers than also delegate to defaultBearerTokenResolver
		// to see if it resolves there as well. For example, if it resolves more than once than throw exception as per spec
     
	}
}
```

As far as configuration goes, let's assume a user would configure a custom `BearerTokenResolver` via `httpSecurity.oauth2().resourceServer().tokenResolver()`.

The `SecurityConfigurer` will always create a `DelegatingBearerTokenResolver` and configure it with the one supplied by the user as an additional resolver. The main point here is that the `SecurityConfigurer` will always create a `DelegatingBearerTokenResolver` and therefore the default will always be `AuthorizationHeaderBearerTokenResolver`.

The delegating strategy is a common design pattern that is already used throughout the framework and provides great flexibility and adaptability in it's design.

vpavic replied:
Thanks for the clarification @jgrandja. The delegating resolver approach would've been valid if the bearer token transport methods were fully exclusive.

I'll emphasize some aspects of RFC 6750 from my previous comment one more time:

- https://tools.ietf.org/html/rfc6750#section-2:
> This section defines three methods of sending bearer access tokens in resource requests to resource servers. **Clients MUST NOT use more than one method to transmit the token in each request.**

- https://tools.ietf.org/html/rfc6750#section-3.1:
> invalid_request
The request is missing a required parameter, includes an unsupported parameter or parameter value, repeats the same parameter, **uses more than one method for including an access token**, or is otherwise malformed.  The resource server SHOULD respond with the HTTP 400 (Bad Request) status code.

The pseudo-code presented above not only violates this by completely ignoring presence of optional methods by default and thus not detecting use of multiple methods, but it also actually prefers the optional method(s) to the required one. To satisfy the requirements quote above, the server needs to parse all 3 methods anyway and at that point it's only about what to do with them.

jgrandja replied:
@vpavic 

> The pseudo-code presented above...

This is strictly pseudo-code and has not been fully thought out so please don't take it as a literal implementation. My purpose was to flush out the design.

Again, as mentioned, the design presented is a common pattern that is being used throughout the framework and is quite flexible as far as plugging in other implementations and gives the "delegating" construct the ability of a "controller". 

> The delegating resolver approach would've been valid if the bearer token transport methods were fully exclusive.

The implementation of `DelegatingBearerTokenResolver` will work and will fulfil the spec requirements. Sounds like you don't agree?

> To satisfy the requirements quoted above, the server needs to parse all 3 methods anyway and at that point it's only about what to do with them.

Why can't the `DelegatingBearerTokenResolver` parse all internal resolvers to determine if more than one method is used by the client? It acts as a controller and it therefore has this control/ability to do so.

vpavic replied:
> The implementation of `DelegatingBearerTokenResolver` will work and will fulfil the spec requirements. Sounds like you don't agree?

From the approach you're describing I don't believe that's not the case.

A request that presents Bearer Token using multiple transport methods should be treated as invalid regardless of whether Resource Server would actually use methods other than Authorization header. With optional `Set<BearerTokenResolver>` approach you wouldn't be looking at other methods unless appropriate `BearerTokenResolver` is registered.

jgrandja replied:
> A request that presents Bearer Token using multiple transport methods should be treated as invalid

Again, the `optional Set<BearerTokenResolver> ` is pseudo-code and should not be taken literally as an implementation. The focus is on the adaptability/flexibility of the design, implementation needs to get flushed out.

If `DelegatingBearerTokenResolver` had both `AuthorizationHeaderBearerTokenResolver` and `RequestBodyBearerTokenResolver` configured than it can fulfill the spec requirement:

> invalid_request
         The request is missing a required parameter, includes an
         unsupported parameter or parameter value, repeats the same
         parameter, **uses more than one method for including an access
         token**, or is otherwise malformed.  The resource server SHOULD
         respond with the HTTP 400 (Bad Request) status code.

I do not want to provide support for URI Query Parameter method based on a number of Security deficiencies outlined in the spec and that are well-known when transporting sensitive information in URI parameters.

jgrandja replied:
@vpavic As mentioned, @rwinch agrees (during our standup) that the current implementation of `DefaultBearerTokenResolver` is the way to go over the delegating strategy along with providing support for the 3 authentication methods. Main thing is to disable form and query by default but allow for configuration if the user requires it. So basically let's go with what you have.

vpavic commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/resource/BearerTokenErrorCodes.java:
```diff
@@ -0,0 +1,46 @@
+ * @since 5.1
+ * @see <a href="https://tools.ietf.org/html/rfc6750#section-3.1" target="_blank">RFC 6750 Section 3.1: Error Codes</a>
+ */
+public interface BearerTokenErrorCodes {
```
I don't believe that is a good idea. This is a different set of error codes (see [RFC 6750 Section 3.1](https://tools.ietf.org/html/rfc6750#section-3.1)) to one defined in [RFC 6749 Section 4.1.2.1](https://tools.ietf.org/html/rfc6749#section-4.1.2.1) and they should be kept separately.

By extension, this also applies to your remarks on use of `OAuth2Error` and `OAuth2AuthenticationException`.

jgrandja replied:
The `invalid_token` and `insufficient_scope` error codes (new) are still *OAuth 2.0 Errors* and are simply registered in the [OAuth Extensions Error Registration](https://tools.ietf.org/html/rfc6750#section-6.2) as defined in [11.4.  OAuth Extensions Error Registry](https://tools.ietf.org/html/rfc6749#section-11.4).

`invalid_request` or  `OAuth2ErrorCodes.INVALID_REQUEST` can be leveraged for Resource Server as well rather than duplicating this error code for Resource Server.

jzheaux replied:
Joe, does being registered in the OAuth Extensions Error Registration promote those error codes outside of the context in which they were defined? For example, it seems to me that it would be inappropriate to use `insufficient_scope` outside of a Resource Server context. (correct me if I'm wrong here)

This would be my only concern with adding bearer-token-focused codes to an umbrella class, that being misuse in other scenarios.

What are your thoughts on maybe having `BearerTokenErrorCodes extends OAuth2ErrorCodes` and `BearerTokenError extends OAuth2Error`? 

As for simply using `OAuth2Error`, the error details for `insufficient_scope` say

> The request requires higher privileges than provided by the
         access token.  The resource server SHOULD respond with the HTTP
         403 (Forbidden) status code and MAY include the "scope"
         attribute with the scope necessary to access the protected
         resource.

meaning that `BearerTokenError` needs one more member variable than `OAuth2Error`.

vpavic replied:
Regardless of the fact OAuth Extensions Error Registration, they are still scoped by their location of use and shouldn't be used interchangeably everywhere. By stuffing everything to `OAuth2ErrorCodes` (and reusing `OAuth2Error` and `OAuth2AuthenticationException`) we exactly encourage that.

IMO Nimubus SDK gets this right by having `com.nimbusds.oauth2.sdk.OAuth2Error` and `com.nimbusds.oauth2.sdk.token.BearerTokenError`.

Besides, we talked the other day how parts of Resource Server (namely filter and everything bearer token resolution related) has the potential of being used outside of OAuth2 flow, i.e. with bearer token of origin other than OAuth2. We can forget that if we start coupling this to core OAuth2 related concerns.

jgrandja replied:
I was anticipating the use of `invalid_token` and `insufficient_scope` outside of the Resource Server context but the more I thought about it the more I couldn't find a use case. So agree that we should not add these to `OAuth2ErrorCodes`.

I still feel that `BearerTokenAuthenticationException` and `BearerTokenError` are redundant. `BearerTokenAuthenticationException` is the same as `OAuth2AuthenticationException`. However, `OAuth2Error` is very similar to `BearerTokenError` but it adds that extra member `scope`. Do we really need this extra member? Why can't the `scope(s)` be specified in the `message` of `OAuth2AuthenticationException` when constructed? The way I see it happening is that the `AuthenticationProvider` or `AccessDecisionVoter` will ultimately create an `OAuth2AuthenticationException` with the following `OAuth2Error`:

`oauth2Error.errorCode` = "insufficient_scope"
`oauth2Error.description` = "The following scopes are required to access this resource -> scope1, scope2"

jgrandja replied:
Just to point out as well, `OAuth2Error` was specifically placed in `oauth2-core` and meant to be reused across the 3 roles/modules. It represents an OAuth 2.0 Error regardless of it's context. However, agree that context specific `ErrorCodes` should be placed in it's appropriate module.

vpavic replied:
The exception specific to Bearer Token authentication is needed to ensure the accompanying implementation of `AuthenticationEntryPoint` (which is the subject of `spring-projects/spring-security#5125` that depends on what we're doing here) is guaranteed to return `WWW-Authenticate` header with only error codes allowed by the spec in https://tools.ietf.org/html/rfc6750#section-3.1.

If you'd use generic `OAuth2AuthenticationException` how do you guard from `WWW-Authenticate` being populated with error codes that shouldn't be used in this context?

jgrandja replied:
The `AuthenticationEntryPoint` implementation would accept an `OAuth2AuthenticationException` and it's associated `OAuth2Error`. One of the evaluations in the implementation would be:

```
if OAuth2Error.errorCode.equals("invalid_token")
... then send back 401 and WWW-Authenticate...
```

I don't see any issues here. If you still do, please be specific on the use case you are pointing out so I can answer it more effectively.

jgrandja replied:
@vpavic As discussed, let's keep `BearerTokenAuthenticationException` but would like you to consider `BearerTokenError` to `extend OAuth2Error` to leverage the 3 common attributes.

vpavic replied:
Now that I got back to this, I recalled why I haven't extended the `OAuth2Error` initially - it's marked as `final`.

jgrandja replied:
I'll remove the `final` modifier and commit/push so you can leverage it.

vpavic replied:
Thanks, sounds good.

vpavic commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/resource/web/BearerTokenResolver.java:
```diff
@@ -0,0 +1,37 @@
+ * @since 5.1
+ * @see <a href="https://tools.ietf.org/html/rfc6750#section-2" target="_blank">RFC 6750 Section 2: Authenticated Requests</a>
+ */
+public interface BearerTokenResolver {
```
Authorization Server module that needs this code also serves in role of Resource Server (for example, OpenID Provider with UserInfo Endpoint) and should pull in `oauth2-resource-server` module. If we follow the logic to move things under `oauth2-core` because Authorization Server could use it then all of the Resource Server code might as well go into `oauth2-core`.

jgrandja replied:
Agreed - `oauth2-authorization-server` will pull in `oauth2-resource-server`

vpavic commented:
I've updated the PR with the following changes:

- addressed the concerns from discussion on this PR (basically this boils down to having `BearerTokenError` extend `OAuth2Error`)
- rebased on top of current `master` to pick up changes from #6
- wired `BearerTokenResolver` to `BearerTokenAuthenticationFilter`

I've also pushed a [WIP branch](https://github.com/vpavic/spring-security-oauth2-resource-server/tree/gh-5125) for spring-projects/spring-security#5125 to preview `AuthenticationEntryPoint` implementation that leverages work from this PR.

jzheaux commented:
I'm happy. @jgrandja?

jgrandja commented:
Yes, looks good.

jzheaux commented:
@vpavic, merged. Thanks, again, for your contributions.
